Capture primitive data type arguments for profiling python_function

shengfukevin opened the pull request:
RECORD_FUNCTION in python_function only captures argument that is a Tensor. However, it is very common for user to use non tensor arguments in custom ops, for example, sequence length in GPT attention custom op. My previous PR tries to capture all non-tensor arguments, it turned out in some cases, it is very expensive.

This PR is to support primitive (or its container) arguments in RECORD_FUNCTION.

soulitzer commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -496,17 +496,43 @@ inline InferredType tryToInferType(py::handle input) {
+  }
+
+  // Try container types
+  return tryToInferContainerType(input, true);
```
We should make sure this prints a nice error message that the user can use to figure out their problem

shengfukevin replied:
I added a warning in python_function.cpp if the data type is not supported.

soulitzer commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -496,17 +496,43 @@ inline InferredType tryToInferType(py::handle input) {
 
-inline InferredType tryToInferContainerType(py::handle input) {
+// This function is similar to tryToInferType, but it only tries to infer
+// primitive types (int, float, bool, complex).
```
Can we document this in the profiler docs for this feature?

shengfukevin replied:
Can you send me the link to the profiler docs?

shengfukevin replied:
I will update https://www.internalfb.com/intern/wiki/PyTorch/Using_PyTorch/PyTorch_Execution_Trace/ after PR is landed

soulitzer reviewed (commented):
Thanks for the quick update, some minor comments

soulitzer commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -889,12 +895,28 @@ std::pair<UnpackedInput, InputFlags> unpack_input(PyObject* args) {
+          PyErr_WarnEx(
+              PyExc_UserWarning,
+              "Only primitive type (bool, float, int, complex) or its container (list, tuple, dict) is supported for non-tensor input in python function profiler",
+              0);
```
I'm not sure we want this warning? It seems that in most usages that go through this path, we won't actually need the record_function_inputs to be recorded. The only case is where we want to do some kind of replay.

I think you should either have an explicit mode which you can enable that says I need to replay later, and if that mode is enabled, you should error immediately here.

The other option is to not do anything here, and just error when you actually go and do the replay.

shengfukevin replied:
I removed it. It is not straight forward to know if user want to capture execution trace. @briancoutinho is working on combining Kineto trace and execution trace. After that change is in, it should be trivial to check it. I will update  profiler_need_input to profiler_need_replay after Brian's change.

soulitzer commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -496,17 +496,43 @@ inline InferredType tryToInferType(py::handle input) {
 
-inline InferredType tryToInferContainerType(py::handle input) {
+// This function is similar to tryToInferType, but it only tries to infer
+// primitive types (int, float, bool, complex).
```
or container of primitive types, and nesting seems to be allowed?

shengfukevin replied:
fixed.

soulitzer commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -590,16 +622,26 @@ inline InferredType tryToInferContainerType(py::handle input) {
-        "."));
+    if (primitiveTypeOnly) {
+      return InferredType(c10::str(
+          "Only tuple, list, or dict of primitive types (bool, float, int, complex)",
```
(possibly nested?)

shengfukevin replied:
fixed.

shengfukevin commented:
@soulitzer , I addressed your code review comments, please take a look. Thanks!

soulitzer reviewed (approved):
Looks good, please double check the perf again before landing

awgu commented:
The CPU overhead issue looks resolved now in my use case! Thanks!

shengfukevin committed:
Capture primitive data type arguments for profiling python_function (#120949)

Summary:
RECORD_FUNCTION in python_function only captures argument that is a Tensor. However, it is very common for user to use non tensor arguments in custom ops, for example, sequence length in GPT attention custom op. My previous PR tries to capture all non-tensor arguments, it turned out in some cases, it is very expensive.

This PR is to support primitive (or its container) arguments in RECORD_FUNCTION.

bypass-github-pytorch-ci-checks


Test Plan:
unit test
    buck test  mode/dev-nosan caffe2/test:profiler -- test_execution_trace_with_pt2 test_execution_trace_alone test_execution_trace_with_kineto test_execution_trace_start_stop test_execution_trace_repeat_in_loop test_execution_trace_no_capture

Reviewed By: soulitzer

Differential Revision: D54525618

Pulled By: shengfukevin
